### ISSUE TITLE
firstOption returning Some(null) (Slick 3.0.3)

### ISSUE THREAD

**Issue by dk8996** [slick/slick#1373]
We ran in to this problem using `firstOption` on version 3.0.3 of slick.  You can find the query with `GROUP_CONAT` below.   If `GROUP_CONAT` is empty `firstOption` returns `Some(null)`.   Here is a possible fix for this, we should use `Option(x)` instead of `Some(x)` since `Option(null)` will be mapped to `None` while `Some(null)` will be mapped to `Some(null)`.

Possible Fix:

```
  final def firstOption(implicit session: JdbcBackend#Session): Option[R] = {
    var res: Option[R] = None
    foreach({ x => res = Option(x) }, 1)
    res
  }
```

Query:

```
          val query = sql"""SELECT GROUP_CONCAT(int_twitter_page_id)
                           FROM page 
                           WHERE page_id IN (#$cs)"""
                val compSet = DB.db.withSession {
                    implicit session =>
                        query.as[String].firstOption
                }
```

(code quoting fixed by @cvogt 2016-07-11)

**Comment by ShaneDelmore** [comment]
Just ran into this today, headOption also returns Some(null) instead of null in this situation.

**Comment by arussel** [comment]
Postgres has some function that would return null in some situation such as array length would return null if you don't have array of the right dimension, so we have the result: 
None if now rows are returned
Some(null) if rows are returned but no array could be 'lengthed' because of arrays dim
Some(x) where x in an Int

We would not want Some(null) to be None in that case. To work with slick, we either coalesce in sql"" or use .?.getOrElse(xx) giving in this case the default value we have.

**Comment by radsaggi** [comment]
@dk8996 Could you please link us to some sample code where you are facing problems? It would definitely be very helpful.
Interestingly this is related to my work [here](https://github.com/slick/slick/pull/1516). We have been discussing some things along those lines in that PR.
@cvogt I guess I could have a look since I am already getting along with the [headOption issue](https://github.com/slick/slick/issues/711)

**Comment by radsaggi** [comment]
@arussel I guess the `GROUP_CONCAT` problem calls for a separate issue, since it looks like it might need to be tackled in a completely independent manner.
@cvogt suggestions?

**Comment by dk8996** [comment]
@radsaggi there is some sample code in my original post.  Re-posting it here:

`val query = sql"""SELECT GROUP_CONCAT(int_twitter_page_id) FROM page WHERE page_id IN (#$cs)""" 

val compSet = DB.db.withSession { implicit session => query.as[String].firstOption }`

**Comment by cvogt** [comment]
Actually Some(null) seems like the correct response here. Yes there was a record, so Some, and yes the value was null, so Some(null).

What happens if you do

```
val query = sql"""SELECT GROUP_CONCAT(int_twitter_page_id) FROM page WHERE page_id IN (#$cs)"""
val compSet = DB.db.withSession { implicit session => query.as[Option[String]].firstOption }
```

?  (See result type in `as`)

Do you get `Some(None)` ?

**Comment by radsaggi** [comment]
@dk8996 May you please verify the code as suggested by @cvogt above so that we may mark this as resolved.

**Comment by radsaggi** [comment]
Marking this as resolved - Working as Intended.

**Comment by dk8996** [comment]
@cvogt the code you provided seems to work, with the addition of `flatten`.  @radsaggi I still think the design for `firstOption` is incorrect.  See discussions about returning `Some(null)`, links below.

[discussion 1](http://www.scala-archive.org/Option-T-and-null-td2008053.html) 
[discussion 2](http://www.scala-lang.org/old/node/6464)

**Comment by cvogt** [comment]
Right, so my example actually returned `Option[Option[String]]`. None meaning there was no row. Some(None) meaning, there was a row and the value was None. Some(Some( "foo" )) meaning there was a row and the value was "foo". Using `.flatten` merges the first two cases into one.

If your column is not Option type, you tell Slick not do do anything about nulls, because what could it do? The only options are return `null` or crash. 

The operations .headOption and .firstOption are using Option to signal if there was a row or not. NOT to identify the value was null. For that you'd need to use the column type.

So in summary it still seems to me that:

> Actually Some(null) seems like the correct response here. Yes there was a record, so Some, and yes the value was null, so Some(null).

And that you'd have to use an Option type if you want Slick to handle that for you.

Chris

Feel free to re-open if you want to make a case for a different view.
